ENH: Add custom descriptors (such as dtype, nunique, etc.) to Styler output

attack68 opened the pull request:
- [x] closes: #43875 


![Screen Shot 2021-10-24 at 08 52 28](https://user-images.githubusercontent.com/24256554/138583994-6a8debd6-6b8a-4fbb-a697-536b5a6b3590.png)

<details>
<summary>Former example...</summary>

![Screen Shot 2021-10-05 at 23 37 31](https://user-images.githubusercontent.com/24256554/136106494-c770db25-5598-4a6b-8708-1d2654eac8e1.png)
</details>

attack68 committed:
ignore hidden rows in loop

attack68 committed:
add latex 43644 test

attack68 committed:
add latex 43644 test

attack68 committed:
Merge remote-tracking branch 'upstream/master' into bug_styler_multiindex_hiding

# Conflicts:
#	pandas/tests/io/formats/style/test_to_latex.py

attack68 committed:
clean up code

attack68 committed:
clean up code

attack68 committed:
row, col and level css

attack68 committed:
whats new

attack68 committed:
tests and user guide

attack68 committed:
merge

attack68 committed:
merge

attack68 committed:
move private methods

attack68 committed:
Merge branch 'clean_styler_css_classes' into bug_styler_multiindex_hiding

# Conflicts:
#	pandas/io/formats/style_render.py
#	pandas/tests/io/formats/style/test_html.py

attack68 committed:
refactor methods

attack68 committed:
add test for 43703

attack68 committed:
add test for 43703

attack68 committed:
docs

attack68 committed:
docs

attack68 committed:
more explicit test

attack68 committed:
more explicit test

attack68 committed:
fix checks

attack68 committed:
fix checks

attack68 committed:
fix checks

attack68 committed:
Merge remote-tracking branch 'upstream/master' into bug_styler_multiindex_hiding

attack68 committed:
fix checks

attack68 committed:
Merge remote-tracking branch 'upstream/master' into clean_styler_css_classes

attack68 committed:
fix checks

attack68 committed:
fix checks

attack68 committed:
refactor to get tests to pass

attack68 committed:
refactor to get tests to pass

attack68 committed:
Merge branch 'clean_styler_css_classes' into bug_styler_multiindex_hiding

attack68 committed:
try filter

attack68 committed:
docs methods

attack68 committed:
correct css classes in set_table_styles

attack68 committed:
Merge remote-tracking branch 'upstream/master' into clean_styler_css_classes

attack68 committed:
add version added

attack68 committed:
add version added

attack68 committed:
checks fix

attack68 committed:
add tests allow none

attack68 committed:
fix checks

attack68 committed:
Merge remote-tracking branch 'upstream/master' into clean_styler_css_classes

attack68 committed:
rename `css` `css_class_names`

attack68 committed:
rename `css` `css_class_names`

attack68 committed:
rename `css` `css_class_names`

attack68 committed:
update arg name

attack68 committed:
Merge remote-tracking branch 'upstream/master' into clean_styler_css_classes

attack68 committed:
fix line length

attack68 committed:
black

attack68 committed:
add vars

attack68 committed:
Merge branch 'clean_styler_css_classes' into bug_styler_multiindex_hiding

# Conflicts:
#	pandas/tests/io/formats/style/test_style.py

attack68 committed:
Merge branch 'bug_styler_multiindex_hiding' into describe_styler

attack68 committed:
add descriptor rendering

attack68 committed:
mypy

pep8speaks commented:
Hello @attack68! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/io/formats/style_render.py`](https://github.com/pandas-dev/pandas/blob/e5b4d3e0ba31969c2529491655a5a59f8f510fe0/pandas/io/formats/style_render.py):

> [Line 509:17](https://github.com/pandas-dev/pandas/blob/e5b4d3e0ba31969c2529491655a5a59f8f510fe0/pandas/io/formats/style_render.py#L509): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'

attack68 committed:
fix merge

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

attack68 committed:
docs

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

attack68 committed:
tests edit

attack68 committed:
some tests

attack68 committed:
doc string

attack68 committed:
add direct test

attack68 committed:
doc fixes

attack68 committed:
update user guide

attack68 committed:
table styles doc

attack68 committed:
version identifiers

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

attack68 committed:
merge master

attack68 committed:
add tests

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

attack68 committed:
whats new

attack68 commented:
@jreback what do you think here? I think this is quite a useful feature for data exploration, and I can see how users might adapt to using it rather than the more generic `df.describe()` which often doesn't contain what you want and has some stats you don't need.
![Screen Shot 2021-11-16 at 18 24 27](https://user-images.githubusercontent.com/24256554/142034998-cd76dea6-2ed1-4f8e-a057-0bbd5e31db17.png)

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

# Conflicts:
#	doc/source/whatsnew/v1.4.0.rst

attack68 committed:
Merge branch 'master' into describe_styler

attack68 committed:
formatting

attack68 committed:
fdoc string validate

jreback commented:
I c so this is creating *arbitrary* summarizations actually.

jreback commented:
cc @pandas-dev/pandas-core any thoughts here.

mroeschke commented on pandas/io/formats/style.py:
```diff
@@ -2217,6 +2231,46 @@ def set_table_styles(
             self.table_styles = table_styles
         return self
 
+    def set_descriptors(
```
Is it implied that this would just be for the header or would `set_header_descriptors` be clearer?

attack68 replied:
i think there is currently only one optional place where it cab go at the moment, and hopefully would be clear from examples. set_descriptors leaves the door open for a footer=False option to move the descriptor to column bottom. might be useful for subtotals?

mroeschke commented on pandas/io/formats/style_render.py:
```diff
@@ -450,6 +461,105 @@ def _generate_col_header_row(self, iter: tuple, max_cols: int, col_lengths: dict
+        ] * (self.index.nlevels - sum(self.hide_index_) - 1)
+
+        if isinstance(descriptor, str):
+            name: str | None = descriptor
```
What happens if the method name doesn't reduce? e.g. `describe`?

attack68 replied:
i think in that case the rendered cell output is the str representation of the result. I.e you can render a dataframe to the cell, albeit it might not look pretty.

attack68 replied:
docs do say func should reduce to scalar, however

mroeschke commented on pandas/io/formats/style_render.py:
```diff
@@ -450,6 +461,105 @@ def _generate_col_header_row(self, iter: tuple, max_cols: int, col_lengths: dict
+                visible_col_count += 1
+                try:
+                    header_element_value = display_func(self.data[col])
+                except Exception:
```
Why not be strict here and raise?

attack68 replied:
in the interest of a simpler initial PR i did not include a subset argument so the function needs to work on all colums. some dtypes might bot be suitable for the function so best approach is silent raise i think. descriptors aim to be more helpful than precise, so i think raising when failing would be annoying

mroeschke replied:
I see. What you think about raising a `UserWarning` that applying `func` failed for a column? Would that be too noisy?

attack68 replied:
I still think so, but even if it wasn't, the way to turn it off is difficult:
  - either a pandas global option is created: I think that's a bit messy for such a small thing.
  - or since the descriptors are calc'ed at render time a `show_warnings` kwarg would have to be saved in `Styler` namespace and evaluated upon render: still a bit messy.

However, with you prompting me to think about it, probably means better documentation and better examples could highlight this and make the behaviour (and workarounds) clear.

bashtage commented:
Feels like it should be aligned with `describe` a lot more closely.  The implementation feels quite clunky, hard to discover, and requiring a relatively high amount of code (e.g., with multiple lambdas to get a few rows).  Why not let rows from `describe` be prepended rather than relying exclusively on UDFs?

attack68 commented:
> Feels like it should be aligned with `describe` a lot more closely. The implementation feels quite clunky, hard to discover, and requiring a relatively high amount of code (e.g., with multiple lambdas to get a few rows). Why not let rows from `describe` be prepended rather than relying exclusively on UDFs?

I'd suggest some of the described clunkiness might be fitting the calculated values into the rendering process/dict that Styler uses to parse. Even if values were taken from `describe` they would still need to be fit into this structure similarly?

The core part of the requests/issue put forward here is that `describe` contains either more info than wanted or not enough, and you can't customise it. And since Styler is a visualisation tool you want the results to display consistently with the other Styler settings.

If you are suggesting expanding `describe` to be more customisable, and then reference that from Styler, that could work, if we pass through the kwargs and retrieve the values and then render them. But that will take more time and effort to go through all review processes, imo.

bashtage commented:
> I'd suggest some of the described clunkiness might be fitting the calculated values into the rendering process/dict that Styler uses to parse. Even if values were taken from describe they would still need to be fit into this structure similarly?

It would be a more user friendly API if users could pass the summary measures by name for any remotely common name, e.g., [`"mean", "std dev", "skewness", "kurtosis"] is more user friendly than writing and passing 4 lambdas. 

> The core part of the requests/issue put forward here is that describe contains either more info than wanted or not enough, and you can't customise it.

Maybe the solution is to customize `describe` and then to use the customized `describe` prepended on a `DataFrame`?

> If you are suggesting expanding describe to be more customisable, and then reference that from Styler, that could work, if we pass through the kwargs and retrieve the values and then render them. But that will take more time and effort to go through all review processes, imo.

Ultimately this is just a prepender to a styled `DataFrame` that applies some particular markup to the prepended values. It should be easy to prepend any second `DataFrame` to an existing `DataFrame` and then to style the output.  That DataFrame could come from `describe` or any other function that returns a `DataFrame` with the correct columns.  Limiting the interface to only lambdas makes it very bespoke to me.

attack68 commented:
> It would be a more user friendly API if users could pass the summary measures by name for any remotely common name, e.g., [`"mean", "std dev", "skewness", "kurtosis"] is more user friendly than writing and passing 4 lambdas.

> Limiting the interface to only lambdas makes it very bespoke to me.

I think you only reviewed one example. Strings, as you describe, are also accepted - you dont have to use lambdas. Shown in docs in original post https://github.com/pandas-dev/pandas/pull/43894#issue-1016904508 and also as response to the issue https://github.com/pandas-dev/pandas/issues/44751#issuecomment-986287632

attack68 commented:
> Ultimately this is just a prepender to a styled `DataFrame` that applies some particular markup to the prepended values. It should be easy to prepend any second `DataFrame` to an existing `DataFrame` and then to style the output. That DataFrame could come from `describe` or any other function that returns a `DataFrame` with the correct columns. Limiting the interface to only lambdas makes it very bespoke to me.

Prepending and styling the result will break all the exsiting functions, like builtin highlighters since there is no way to separate the summary statistics from the data. It would also make the hiding functions more unwieldy, and the formatting might be more difficult too, since columns might lose dtype consistency.

I think potentially using the values from the `describe` DataFrame could work, but only within the Styler rendering process.

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

attack68 commented:
This has stalled, some of the feedback was based on the misunderstanding that one had to use lambdas when this is not true. Appreciate another look.

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

attack68 committed:
push to 1.5.0

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

# Conflicts:
#	pandas/tests/io/formats/style/test_style.py

attack68 committed:
cleaer comment

attack68 committed:
Merge remote-tracking branch 'upstream/master' into describe_styler

# Conflicts:
#	doc/source/whatsnew/v1.5.0.rst

attack68 committed:
fix formatter

attack68 committed:
Merge remote-tracking branch 'upstream/main' into describe_styler

attack68 committed:
fix formatter

attack68 committed:
Merge remote-tracking branch 'upstream/main' into describe_styler

# Conflicts:
#	doc/source/whatsnew/v1.5.0.rst
#	pandas/tests/io/formats/style/test_style.py

attack68 commented:
@mroeschke this stalled a while back. As the original reviewer, care to revive it?

Delengowski commented:
I can't provide a review but I can say this functionality would be very useful. My work currently gets functionality like this through extending the styler with our own subclass and jinja template. This would save us a ton of work.

Really appreciate all the styler work you've done @attack68

mroeschke commented on pandas/io/formats/style_render.py:
```diff
@@ -1418,7 +1532,7 @@ def _default_formatter(x: Any, precision: int, thousands: bool = False) -> Any:
     if isinstance(x, (float, complex)):
         return f"{x:,.{precision}f}" if thousands else f"{x:.{precision}f}"
-    elif isinstance(x, int):
+    elif isinstance(x, (int, np.int64)):
```
Think you can use `from pandas.core.dtypes.common import is_integer`

attack68 replied:
done

mroeschke reviewed (approved):
I don't use styler much, but if the community has an interest in this feature I think it's worthwhile

attack68 committed:
Merge remote-tracking branch 'upstream/main' into describe_styler

attack68 committed:
is_integer replacemnet

attack68 commented:
@jreback what you think here? OK to merge here on green? I can extend the docs as an add on?

jreback commented:
no this still has questions

attack68 commented:
@bashtage I believe the questions jeff is referring to are yours, do you have time to review?

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -124,6 +128,7 @@ def __init__(
         self.hide_columns_: list = [False] * self.columns.nlevels
         self.hidden_rows: Sequence[int] = []  # sequence for specific hidden rows/cols
         self.hidden_columns: Sequence[int] = []
+        self.descriptors: list[str | Callable | tuple[str, Callable]] = []
```
Is this a common API?  I think most places where alternative input are allowed either use a Sequence of some sort or `dict[str, str | Callable]`.  Would be be cleaner if the allowed inputs were `Sequence[str | Callable] | dict[str | Callable]`.  

Also, can't the `Callable be better described?  Isn't is `Callable[[Series], float]`?  It seems that the `Callable` must take a `Series` and return some sort of scalar value.  Thie question is whether the scalar value can be non-numeric.  If it could be any numeric value, than you could you use something like `Callable[[Series], int | float]`.  So more generally, what are the requirements for the `Callable`?

attack68 replied:
> 
> Also, can't the `Callable be better described? Isn't is `Callable[[Series], float]`? It seems that the `Callable`must take a`Series`and return some sort of scalar value. Thie question is whether the scalar value can be non-numeric. If it could be any numeric value, than you could you use something like`Callable[[Series], int | float]`. So more generally, what are the requirements for the `Callable`?

The Callable could techincally return anything so long as the returned object has a `__str__` representation to populate in the HTML element. Common values might be int, float, string, but yes it does accept only series.

bashtage replied:
I think the sig should then be `Callable[[Series],Any]`

attack68 replied:
- [x] changed typing. But did not add the dict. `List[Tuple[str, str]]` is present for some of the `Styler` functions

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -365,7 +372,12 @@ def _translate_header(self, sparsify_cols: bool, max_cols: int):
-        # 2) index names
+        # 2) Descriptor calcs
+        for r, descriptor in enumerate(self.descriptors):
+            descriptor_row = self._generate_descriptor_row((r, descriptor), max_cols)
```
Why pass a tuple to this function?  It would be better to pass the row and the descriptor as separate arguments since there is no gain to repacking them, since you immediately unpack them in the function.

attack68 replied:
agreed, but all the other generator methods use this format (rightly or wrongly) so I stuck to the same format for consistency. is minor either way.

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -477,6 +489,109 @@ def _generate_col_header_row(self, iter: tuple, max_cols: int, col_lengths: dict
+        elif isinstance(descriptor, tuple):
+            name, func = descriptor[0], descriptor[1]
+        else:
+            name, func = None, descriptor
```
When the descriptor is a `Callable` I think it would be best to try `getattr(descriptor, "name", None)` for the `name` rather than committing to a blank name. For example `pd.Series.mean.__name__` is `mean`.

attack68 replied:
- [x] chged. note the case of `"<lambda>"` i thought was worth blanking

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -477,6 +489,109 @@ def _generate_col_header_row(self, iter: tuple, max_cols: int, col_lengths: dict
+            name, func = None, descriptor
+
+        display_func: Callable = _maybe_wrap_formatter(
+            formatter=None,  # use _default_formatter
```
Default value is already `None`.

attack68 replied:
- [x] chged

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -477,6 +489,109 @@ def _generate_col_header_row(self, iter: tuple, max_cols: int, col_lengths: dict
+            _element(
+                "th",
+                base_css
+                if (name is not None and not self.hide_column_names)
```
This is awfully hard to read code with 2 ternary operators inside a function inside a list.

bashtage replied:
You could defer the list packing until the end.

attack68 replied:
- [x] chged

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -1433,7 +1548,7 @@ def _wrap_decimal_thousands(
 
     def wrapper(x):
-        if isinstance(x, (float, complex, int)):
+        if isinstance(x, (float, complex)) or is_integer(x):
```
For consistency you should use `is_float` and `is_complex` which are looser (e.g., allowing NumPy float64)

attack68 replied:
- [x] chged

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -124,6 +128,7 @@ def __init__(
         self.hide_columns_: list = [False] * self.columns.nlevels
         self.hidden_rows: Sequence[int] = []  # sequence for specific hidden rows/cols
         self.hidden_columns: Sequence[int] = []
+        self.descriptors: list[str | Callable | tuple[str, Callable]] = []
```
If you do stick with the tuple, it should be `tuple(str, str | Callable)` which would allow someone to pass a custom name for the output along with a common function, e.g., `("average", "mean").  If moving to a dict, then both `{"average":"mean"}` and `{"average":Series.mean}` should be allowed.

attack68 replied:
yes, agreed, having a think if this can be improved.

bashtage reviewed (changes requested):
I think it is a lot more useful now.  The biggest question in this review is what the more API should look like.  It seems to be that allowing either `Sequence[str | Callable]` or `dict[str, str | Callable]` is more consistent with other parts of pandas that are flexible.  I can't think of any cases where `tuple(str, str | Callable)` would be used to get a string name and it's associated argument.

attack68 committed:
Merge remote-tracking branch 'upstream/main' into describe_styler

attack68 committed:
bastage req: refactor _element

attack68 committed:
bashtage req: is_float / complex

attack68 committed:
bashtage req: is_float / complex

attack68 committed:
bashtage req: __name__ and typing

attack68 committed:
bashtage req: doc updates for __name__

attack68 committed:
doc fix

jreback commented:
I am still -1 here. This is basically completely changing the way things work. We *already* have .describe() to do exactly this. If you want to add `.describe()` and then allow it to be more generic ok there. BUT then this adds these as header summary rows, again completely changing things.

I don't have a good recommendation for how to reconcile this. I support if allow a 'header' table would be more ameanable. This just feels tacked on with a lot of bespoke things going on.

Delengowski commented:
> I am still -1 here. This is basically completely changing the way things work. We _already_ have .describe() to do exactly this. If you want to add `.describe()` and then allow it to be more generic ok there. BUT then this adds these as header summary rows, again completely changing things.
> 
> I don't have a good recommendation for how to reconcile this. I support if allow a 'header' table would be more ameanable. This just feels tacked on with a lot of bespoke things going on.

I think I get what you are saying. I think generalizing `.describe()` would be a smart move, but keep this api, have the actual stylization call the new `.describe()`. I'm of the opinion that the addition of the data descriptors to the header does belong on `styler`, but I get your concern of having this _extra_ describe functionality be under `styler` rather than under `.describe()` (if that's indeed what you are talking about).

What I don't think should been done is have these generalized functionalities moved to `.describe()` and then have api be that the end user call `.describe()`  and be forced to pass the resultant  `DataFrame`/`Series`  to some kwarg on `styler`.

attack68 commented:
@bashtage @jreback  @mroeschke, this PR stalled due to difficulty in finding a way forward. I have tried to refactor it addressing some concerns above in an alternate #46105. If you have time please comment on the alternative.

Delengowski commented:
I'd be interested in hearing a more detailed response to the opposition of this. The styler is an end use feature, so I'm not sure I understand the issue with there being data descriptors in the header or in a separate footer. 

I understand and agree to the issue with duplicating computation logic but not with adding extra rows to the header. Again it's end use, where further computation isnt the primary purpose.

jreback commented:
api considerations are always paramount - the goal is consistency and simplicity 
not convering every possible case

will look soon

mroeschke commented:
There is a Separation of Concerns argument here: 1) the part to generate the descriptor data & combine with the original data, 2) the part to stylize the result. Without this feature today it's maybe roughly equivalent to (?):

```
result = pd.concat([original_data.astype(object), descriptor.astype(object)])
output = result.style.(...)
```

Maybe there should be a better way to do 1 (that's why `describe()` was brought up, `pivot_table` has similar functionality)

attack68 commented:
@mroeschke solution, with the concatenation, might work but it then becomes harder to index the original data, export the styler for more reuseable styling, and apply the builtin formatters since you then have to exclude the concatenated rows from your subset.

I have used his solution before for a "Total" row and it is cumbersome but worked.

bashtage commented on pandas/io/formats/style.py:
```diff
@@ -2423,6 +2440,50 @@ def set_table_styles(
+        Parameters
+        ----------
+        descriptors : list of str, callables or 2-tuples of str and callable
+            If a string is given must be a valid Series method, e.g. "mean" invokes
```
Could you some additional details about the callable.  Specifically it must (or should) return a scalar quantity.  Is the callable allowed to throw and exception?  If so, what happens?  I would be good to clarify this.

bashtage commented on pandas/io/formats/style.py:
```diff
@@ -2423,6 +2440,50 @@ def set_table_styles(
+        ...     return s.mean()
+        >>> styler = df.style.set_descriptors([
+        ...     "mean",
+        ...     Series.mean,
```
The screen show shows no lable on this line? Was it not possible to have `.__name__` as a default test?

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -124,6 +133,7 @@ def __init__(
         self.hide_columns_: list = [False] * self.columns.nlevels
         self.hidden_rows: Sequence[int] = []  # sequence for specific hidden rows/cols
         self.hidden_columns: Sequence[int] = []
+        self.descriptors: list[Descriptor | tuple[str, Descriptor]] = []
```
Why not make simplify this to `self.descriptors: list[tuple[str, Descriptor]]`?  Also, should it immutable, so that it would be `self.descriptors: tuple[tuple[str, Descriptor], ...]`?

Essentially if the user doesn't provide a name, then the name used internally, even if blank, is appended so that the tuple form is the only one that would ever need to be used.

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -477,6 +494,108 @@ def _generate_col_header_row(self, iter: tuple, max_cols: int, col_lengths: dict
 
         return index_blanks + column_name + column_headers
 
+    def _generate_descriptor_row(self, iter: tuple, max_cols: int):
```
What kind of tuples are allowed? Would be better to be as specific as possible to reduce any future refactor risks.

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -477,6 +494,108 @@ def _generate_col_header_row(self, iter: tuple, max_cols: int, col_lengths: dict
+            else:
+                func = descriptor[1]
+        else:
+            name, func = getattr(descriptor, "__name__", None), descriptor
```
Should name be `None` of just `""`?  Is there any utility is tracking `None`?  If name is `""` then you can simplify the type above to be `str` rather than `str | None`.

bashtage replied:
Or perhaps just use `self.css["blank_value"]` when missing and simplify below.

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -477,6 +494,108 @@ def _generate_col_header_row(self, iter: tuple, max_cols: int, col_lengths: dict
+                func = descriptor[1]
+        else:
+            name, func = getattr(descriptor, "__name__", None), descriptor
+            name = None if name == "<lambda>" else name  # blank nameless functions
```
Same here.  Replace nameless -> anonymous in comment

bashtage commented on pandas/io/formats/style_render.py:
```diff
@@ -477,6 +494,108 @@ def _generate_col_header_row(self, iter: tuple, max_cols: int, col_lengths: dict
+        )
+
+        base_css = f"{self.css['descriptor_name']} {self.css['descriptor']}{r}"
+        if name is not None and not self.hide_column_names:
```
The above refactor removes this `if` `then` since `name` is always assigned.

bashtage reviewed (changes requested):
Some small refactorings to make it simpler, and perhaps more palatable in the long run.

bashtage commented:
I've some to the point where the LOC is now small enough that it is worth including for the flexibility it brings.  The hard part is always expanding the API as pandas as become more mature, as this requires some effort (although I would rate this as highly stable, and so low maintenance).

Ideally, there would be a way to find out if this feature was in fact widely desirable before committing to it for a fairly long period of time.

attack68 commented:
> I've some to the point where the LOC is now small enough that it is worth including for the flexibility it brings. The hard part is always expanding the API as pandas as become more mature, as this requires some effort (although I would rate this as highly stable, and so low maintenance).
> 
> Ideally, there would be a way to find out if this feature was in fact widely desirable before committing to it for a fairly long period of time.

Thanks @bashtage for the review and I agree with all your points, but I think Jeff was very keen on completely separating the calculation and thereby reducing the LOC much more, which I also agree with, but I was struggling to think of a way.

Finally, I recently had another thought and managed to completely reengineer this with the execution into less than 20 LOC, with better formatting and HTML styling built in, so I'm closing to close this and leave the superior, #46105, open. (thanks also @mroeschke for review)
